tck/src/test/java/org/reactivestreams/tck/RangePublisherTest.java

```diff
@@ -0,0 +1,176 @@
+ * this code has waived all copyright and related or neighboring         *
+ * rights to this code.                                                  *
+ *                                                                       *
+ * You should have received a copy of the CC0 legalcode along with this  *
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.*
+ ************************************************************************/
+
+package org.reactivestreams.tck;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.*;
+
+import org.reactivestreams.*;
+import org.testng.annotations.*;
+
+@Test
+public class RangePublisherTest extends PublisherVerification<Integer> {
+
+    static final Map<Integer, StackTraceElement[]> stacks = new ConcurrentHashMap<Integer, StackTraceElement[]>();
+    
+    static final Map<Integer, Boolean> states = new ConcurrentHashMap<Integer, Boolean>();
+    
+    static final AtomicInteger id = new AtomicInteger();
+    
+    @AfterClass
+    public static void afterClass() {
+        boolean fail = false;
+        StringBuilder b = new StringBuilder();
+        for (Map.Entry<Integer, Boolean> t : states.entrySet()) {
+            if (!t.getValue()) {
+                b.append("\r\n-------------------------------");
+                for (Object o : stacks.get(t.getKey())) {
+                    b.append("\r\nat ").append(o);
+                }
+                fail = true;
+            }
+        }
+        if (fail) {
+            throw new AssertionError("Cancellations were missing:" + b);
+        }
+    }
+    
+    public RangePublisherTest() {
+        super(new TestEnvironment());
+    }
+
+    @Override
+    public Publisher<Integer> createPublisher(long elements) {
+        return new RangePublisher(1, elements);
+    }
+
+    @Override
+    public Publisher<Integer> createFailedPublisher() {
+        return null;
+    }
+
+    static final class RangePublisher
+    implements Publisher<Integer> {
+        
+        final StackTraceElement[] stacktrace;
+
+        final long start;
+        
+        final long count;
+        
+        RangePublisher(long start, long count) {
+            this.stacktrace = Thread.currentThread().getStackTrace();
+            this.start = start;
+            this.count = count;
+        }
+        
+        @Override
+        public void subscribe(Subscriber<? super Integer> s) {
+            if (s == null) {
+                throw new NullPointerException();
+            }
+            
+            int ids = id.incrementAndGet();
+            
+            RangeSubscription parent = new RangeSubscription(s, ids, start, start + count);
+            stacks.put(ids, stacktrace);
+            states.put(ids, false);
+            s.onSubscribe(parent);
+        }
+        
+        static final class RangeSubscription extends AtomicLong implements Subscription {
```

Comment:
The main purpose was to track cancellation/completion in response to the TCK and this was the simplest multi-valued `Publisher` I could think of.

Working with an iterator requires a lot of try-catches around the method calls to `Iterator` because they may throw (for example, I saw libraries that hid JDBC calls behind plain iterators).